### PR TITLE
OpenID Recipes: use step model instead of view model, support update

### PR DESCRIPTION
**Pull request by kaipm**
## Changes in the pull request
- Add unit tests for OpenID scopes
- Add unit tests for OpenID apps
- Replace `OpenIdScopeStepViewModel` with `OpenIdScopeStepModel` in recipe
- Replace `CreateOpenIdApplicationViewModel` with `OpenIdApplicationStepModel` in recipe
- Adjust `OpenIdScopeStep` to support `update`
- Adjust `OpenIdApplicationStep` to support `update`
- Adjust `OpenIdApplicationStep` to include importing scopes, which were not imported before

This is another part of #6364

### CODE REVIEW

**Review by deanmarcussen — COMMENTED** [review]
Are the new step models the same as the old viewmodels that were used?

i.e. is it non breaking on existing recipes, or?

But awesome looks really good thanks

**Comment by sebastienros** [comment]
@kaipm we just need to understand if it is breaking anything so we can document the upgrade steps accordingly.

**Comment by kaipm** [comment]
Yes, they are. I did unit tests before making any changes, to make sure I match the old structure.

However, there is a small difference in the behaviour:

This was necessary to add role1 before the PR, it still works after the PR:
```
"roleEntries": [
  {
    "name": "role1",
    "selected": true
  },
```

The following two examples did **not** add role1 before the PR. But after the PR they **will add** role1 because `selected` is ignored completely:
```
"roleEntries": [
  {
    "name": "role1",
    "selected": false
  },
```
```
"roleEntries": [
  {
    "name": "role1",
  },
```

**Review by deanmarcussen — APPROVED** [review]

File: test/OrchardCore.Tests/Modules/OrchardCore.OpenId/OpenIdApplicationStepTests.cs [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+            appManagerMock.Setup(m =>
+                m.CreateAsync(
+                    It.IsAny<OpenIdApplicationDescriptor>(),
+                    It.IsAny<CancellationToken>()))
+                .Callback<object, CancellationToken>((app, c) =>
+                    actual = (OpenIdApplicationDescriptor)app)
+                .Returns(
+                    new ValueTask<object>(actual));
+
+            var step = new OpenIdApplicationStep(appManagerMock.Object);
```
Comment (hishamco) [inline test/OrchardCore.Tests/Modules/OrchardCore.OpenId/OpenIdApplicationStepTests.cs:55]: Remove extra whitespaces

File: test/OrchardCore.Tests/Modules/OrchardCore.OpenId/OpenIdApplicationStepTests.cs [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+using OrchardCore.Recipes.Models;
+using OrchardCore.Tests.Utilities;
+using Xunit;
+using static OpenIddict.Abstractions.OpenIddictConstants;
+
+namespace OrchardCore.Tests.Modules.OrchardCore.OpenId
+{
+    public class OpenIdApplicationStepTests
+    {
+        private string GetRecipeFileContent(string recipeName)
```
Comment (hishamco) [inline test/OrchardCore.Tests/Modules/OrchardCore.OpenId/OpenIdApplicationStepTests.cs:23]: You can move `private` method before ending the class

**Review by hishamco — CHANGES_REQUESTED** [review]
LGTM, after removing the extra whitespaces

File: src/OrchardCore.Modules/OrchardCore.OpenId/Recipes/OpenIdApplicationStep.cs [hunk 3]
```diff
@@ -29,16 +28,26 @@ public async Task ExecuteAsync(RecipeExecutionContext context)
                 return;
             }
 
-            var model = context.Step.ToObject<CreateOpenIdApplicationViewModel>();
+            var model = context.Step.ToObject<OpenIdApplicationStepModel>();
 
-            var descriptor = new OpenIdApplicationDescriptor
+            var app = await _applicationManager
+                .FindByClientIdAsync(model.ClientId);
+
```
[the remark this answers is not in the document]
Comment (kaipm) [reply src/OrchardCore.Modules/OrchardCore.OpenId/Recipes/OpenIdApplicationStep.cs]: Done

File: src/OrchardCore.Modules/OrchardCore.OpenId/Recipes/OpenIdApplicationStepModel.cs [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+        public bool AllowRefreshTokenFlow { get; set; }
+        public bool AllowHybridFlow { get; set; }
+        public bool AllowImplicitFlow { get; set; }
+        public bool AllowLogoutEndpoint { get; set; }
+
+        public class RoleEntry
+        {
+            public string Name { get; set; }
+        }
+        public class ScopeEntry
```
[the remark this answers is not in the document]
Comment (kaipm) [reply src/OrchardCore.Modules/OrchardCore.OpenId/Recipes/OpenIdApplicationStepModel.cs:29]: Done

File: src/OrchardCore.Modules/OrchardCore.OpenId/Recipes/OpenIdScopeStep.cs [hunk 5]
```diff
@@ -27,20 +26,40 @@ public async Task ExecuteAsync(RecipeExecutionContext context)
                 return;
             }
 
-            var model = context.Step.ToObject<OpenIdScopeStepViewModel>();
-            var descriptor = new OpenIdScopeDescriptor
+            var model = context.Step.ToObject<OpenIdScopeStepModel>();
+
```
[the remark this answers is not in the document]
Comment (kaipm) [reply src/OrchardCore.Modules/OrchardCore.OpenId/Recipes/OpenIdScopeStep.cs]: Done

File: test/OrchardCore.Tests/Modules/OrchardCore.OpenId/OpenIdApplicationStepTests.cs (hunk 1 above)
Comment (kaipm) [reply test/OrchardCore.Tests/Modules/OrchardCore.OpenId/OpenIdApplicationStepTests.cs:47]: Done

File: test/OrchardCore.Tests/Modules/OrchardCore.OpenId/OpenIdApplicationStepTests.cs (hunk 2 above)
Comment (kaipm) [reply test/OrchardCore.Tests/Modules/OrchardCore.OpenId/OpenIdApplicationStepTests.cs]: Done

**Review by hishamco — APPROVED** [review]

**Comment by kaipm** [comment]
@hishamco, I did all the proposed changes.

I also removed some line breaks between IF-statements too, because most were without already. I hope that's fine!

**Comment by hishamco** [comment]
Thanks @kaipm for the PR, we will merge this ASAP

**Comment by hishamco** [comment]
Thanks @kaipm for the PR, we will merge this ASAP

**Pull request merged**